### PR TITLE
Lagrange basis v2

### LINKED ISSUE #9: Top 3
The Lagrange Basis

### PR DESCRIPTION
**Pull request by xc900606**
## Description
This pull request implements the LagrangePolynomial class within the numanalysislib.basis module. It includes the logic for constructing the basis and evaluating it using the Barycentric formula for improved numerical stability.

## Linked Issue
Fixes #9 

## Implementation Checklist
- [x] My class inherits from `PolynomialBasis` (if applicable).
- [x] I have implemented all abstract methods (`evaluate_basis`, `fit`, etc.).
- [x] I have used `numpy` vectorization to avoid inefficient Python loops.
- [x] Documentation strings (docstrings) are provided for all public methods.

### 1. Mathematical Integrity
- [ ] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [ ] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [ ] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [ ] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [ ] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [ ] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [ ] **Coverage:** Unit tests cover typical usage and edge cases.
- [ ] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [ ] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

### CODE REVIEW

File: src/numanalysislib/basis/lagrange.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    def evaluate_basis(self, index: int, x: np.ndarray) -> np.ndarray:
+        """
+        Evaluate the index-th Lagrange basis function at points x.
+        """
+        if self.nodes is None or self.weights is None:
+            raise ValueError("Call fit before evaluating the basis.")
+        if not 0 <= index < self.n_dofs:
+            raise IndexError("Basis index out of range.")
+
+        # 1. Setup
```
Comment (mtezzele) [inline src/numanalysislib/basis/lagrange.py]: no need for comments that do not explain anything

File: src/numanalysislib/basis/lagrange.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        x = np.asarray(x, dtype=float)
+        x_flat = np.atleast_1d(x).reshape(-1)
+
+        # 2. Distance Matrix
```
Comment (mtezzele) [inline src/numanalysislib/basis/lagrange.py]: remove this enumeration

File: src/numanalysislib/basis/lagrange.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            terms = self.weights / diff
+
+            # The denominator is the sum of all terms (same as evaluate)
+            denominator = np.sum(terms, axis=1)
+
+            # The numerator is JUST the term for our specific 'index'
+            numerator = terms[:, index]
+            values = numerator / denominator
+
+        # 4. --- Fix Exact Node Matches (The VIP Rule) ---
```
Comment (mtezzele) [inline src/numanalysislib/basis/lagrange.py]: what's the VIP rule? Please use informative comments and not generative ai ones

File: src/numanalysislib/basis/lagrange.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        # We temporarily ignore "divide by zero" warnings because we will
+        # manually fix exact node matches right after this math.
+        with np.errstate(divide='ignore', invalid='ignore'):
+            # The lambda_j / (x - x_j) piece used in both top and bottom
+            terms = self.weights / diff
+
+            # Numerator: Sum of (terms * y_j)
+            numerator = np.sum(terms * coefficients, axis=1)
+
+            # Denominator: Sum of (terms)
```
Comment (mtezzele) [inline src/numanalysislib/basis/lagrange.py]: why the parenthesis? Also, this is literally what is written in the code, so please remove it.

File: src/numanalysislib/basis/lagrange.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        if np.unique(x_nodes).size != x_nodes.size:
+            raise ValueError("All interpolation nodes must be distinct.")
+        if x_nodes.size != self.n_dofs or y_nodes.size != self.n_dofs:
+            raise ValueError(
+                f"Expected {self.n_dofs} nodes and {self.n_dofs} values."
+            )
+
+        self.nodes = x_nodes
+        self._compute_weights()
+        return y_nodes.copy()
```
Comment (mtezzele) [inline src/numanalysislib/basis/lagrange.py]: can you explain why here you return a copy?

File: src/numanalysislib/basis/lagrange.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+"""
+
+import numpy as np
+from numanalysislib.basis._abstract import PolynomialBasis
+
+
+class LagrangePolynomial(PolynomialBasis):
+    r"""
+    Lagrange polynomial basis defined by a set of distinct
+    interpolation nodes.
```
Comment (BaoTruong1) [inline src/numanalysislib/basis/lagrange.py:23]: You should add what does your function do, what are the parameters, and what does it return to every method.

File: src/numanalysislib/basis/lagrange.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        # --- Fix Exact Node Matches ---
+        # If the user asks to evaluate exactly AT a node (where x - x_j = 0),
+        # the formula breaks (1/0).
+        exact_matches = np.where(diff == 0)
+        if len(exact_matches[0]) > 0:
+            result[exact_matches[0]] = coefficients[exact_matches[1]]
+
+        return result
+
+    def fit(self, x_nodes: np.ndarray, y_nodes: np.ndarray) -> np.ndarray:
```
Comment (BaoTruong1) [inline src/numanalysislib/basis/lagrange.py:147]: This one should return coefficients instead of y_nodes.

File: src/numanalysislib/basis/lagrange.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            result = numerator / denominator
+
+        # --- Fix Exact Node Matches ---
+        # If the user asks to evaluate exactly AT a node (where x - x_j = 0),
+        # the formula breaks (1/0).
+        exact_matches = np.where(diff == 0)
+        if len(exact_matches[0]) > 0:
+            result[exact_matches[0]] = coefficients[exact_matches[1]]
+
+        return result
```
Comment (BaoTruong1) [inline src/numanalysislib/basis/lagrange.py]: Need to reshape the result

**Comment by mtezzele** [comment]
Please fix the conflicts as well

**Comment by mtezzele** [comment]
@xc900606 please work on this PR ASAP, since it is a really important module and it has become a bottleneck

**Comment by xc900606** [comment]
1. removed the unnecessary comments.
2. cleaned up the docstrings
3. fixed the output shape in evaluate_basis and evaluate by reshaping the result back to x.shape.
4. updated fit() to return coefficients explicitly instead of y_nodes.copy(). In the nodal Lagrange basis, the coefficients are the nodal values, so this is clearer
